doc/tinc.8.in
.Dd 2014-01-16
.Dt TINCCTL 8
.\" Manual page created by:
.\" Scott Lamb
.Sh NAME
.Nm tinc
.Nd tinc VPN control
.Sh SYNOPSIS
.Nm
.Op Fl bcn
.Op Fl -config Ns = Ns Ar DIR
.Op Fl -net Ns = Ns Ar NETNAME
.Op Fl -pidfile Ns = Ns Ar FILENAME
.Op Fl -batch
.Op Fl -force
.Op Fl -help
.Op Fl -version
.Op Ar COMMAND
.Sh DESCRIPTION
This is the control program of tinc, a secure virtual private network (VPN)
project.
.Nm
can start and stop
.Xr tincd 8 ,
and can to alter and inspect the state of a running VPN.
It can also be used to change the configuration,
or to import or export host configuration files from other nodes.

If
.Nm
is started with a
.Ar COMMAND ,
this command is immediately executed, after which
.Nm
exits.
If no
.Ar COMMAND
is given,
.Nm
will act as a shell;
it will display a prompt, and commands can be entered on the prompt.
If
.Nm
is compiled with libreadline, history and command completion are available on the prompt.
One can also pipe a script containing commands through
.Nm .
In that case, lines starting with a # symbol will be ignored.
.Sh OPTIONS
.Bl -tag -width indent
.It Fl n, -net Ns = Ns Ar NETNAME
Communicate with tincd(8) connected with
.Ar NETNAME .
.It Fl -pidfile Ns = Ns Ar FILENAME
Use the cookie from
.Ar FILENAME
to authenticate with a running tinc daemon.
If unspecified, the default is
.Pa @runstatedir@/tinc. Ns Ar NETNAME Ns Pa .pid.
.It Fl b, -batch
Don't ask for anything (non-interactive mode).
.It Fl -force
Force some commands to work despite warnings.
.It Fl -help
Display short list of options.
.It Fl -version
Output version information and exit.
.El
.Sh ENVIRONMENT VARIABLES
.Bl -tag -width indent
.It Ev NETNAME
If no netname is specified on the command line with the
.Fl n
option, the value of this environment variable is used.
.El
.Sh COMMANDS
.Bl -tag -width indent
.It init Op Ar name
Create initial configuration files and RSA and Ed25519 key pairs with default length.
If no
.Ar name
for this node is given, it will be asked for.
.It get Ar variable
Print the current value of configuration variable
.Ar variable .
If more than one variable with the same name exists,
the value of each of them will be printed on a separate line.
.It set Ar variable Ar value
Set configuration variable
.Ar variable
to the given
.Ar value .
All previously existing configuration variables with the same name are removed.
To set a variable for a specific host, use the notation
.Ar host Ns Li . Ns Ar variable .
.It add Ar variable Ar value
As above, but without removing any previously existing configuration variables.
If the variable already exists with the given value, nothing happens.
.It del Ar variable Op Ar value
Remove configuration variables with the same name and
.Ar value .
If no
.Ar value
is given, all configuration variables with the same name will be removed.
.It edit Ar filename
Start an editor for the given configuration file.
You do not need to specify the full path to the file.
.It export
Export the host configuration file of the local node to standard output.
.It export-all
Export all host configuration files to standard output.
.It import
Import host configuration data generated by the
.Nm
export command from standard input.
Already existing host configuration files are not overwritten unless the option
.Fl -force
is used.
.It exchange
The same as export followed by import.
.It exchange-all
The same as export-all followed by import.
.It invite Ar name
Prepares an invitation for a new node with the given
.Ar name ,
and prints a short invitation URL that can be used with the join command.
.It join Op Ar URL
Join an existing VPN using an invitation URL created using the invite command.
If no
.Ar URL
is given, it will be read from standard input.
.It start Op tincd options
Start
.Xr tincd 8 ,
optionally with the given extra options.
.It stop
Stop
.Xr tincd 8 .
.It restart Op tincd options
Restart
.Xr tincd 8 ,
optionally with the given extra options.
.It reload
Partially rereads configuration files. Connections to hosts whose host
config files are removed are closed. New outgoing connections specified
in
.Xr tinc.conf 5
will be made.
.It pid
Shows the PID of the currently running
.Xr tincd 8 .
.It generate-keys Op bits
Generate both RSA and Ed25519 key pairs (see below) and exit.
.It generate-ed25519-keys
Generate public/private Ed25519 key pair and exit.
.It generate-rsa-keys Op bits
Generate public/private RSA key pair and exit.
If
.Ar bits
is omitted, the default length will be 2048 bits.
When saving keys to existing files, tinc will not delete the old keys;
you have to remove them manually.
.It dump [reachable] nodes
Dump a list of all known nodes in the VPN.
If the keyword reachable is used, only lists reachable nodes.
.It dump edges
Dump a list of all known connections in the VPN.
.It dump subnets
Dump a list of all known subnets in the VPN.
.It dump connections
Dump a list of all meta connections with ourself.
.It dump graph | digraph
Dump a graph of the VPN in
.Xr dotty 1
format.
Nodes are colored according to their reachability:
red nodes are unreachable, orange nodes are indirectly reachable, green nodes are directly reachable.
Black nodes are either directly or indirectly reachable, but direct reachability has not been tried yet.
.It dump invitations
Dump a list of outstanding invitations.
The filename of the invitation, as well as the name of the node that is being invited is shown for each invitation.
.It info Ar node | subnet | address
Show information about a particular node, subnet or address.
If an address is given, any matching subnet will be shown.
.It purge
Purges all information remembered about unreachable nodes.
.It debug Ar N
Sets debug level to
.Ar N .
.It log Op Ar N
Capture log messages from a running tinc daemon.
An optional debug level can be given that will be applied only for log messages sent to
.Nm tinc .
.It retry
Forces
.Xr tincd 8
to try to connect to all uplinks immediately.
Usually
.Xr tincd 8
attempts to do this itself,
but increases the time it waits between the attempts each time it failed,
and if
.Xr tincd 8
didn't succeed to connect to an uplink the first time after it started,
it defaults to the maximum time of 15 minutes.
.It disconnect Ar NODE
Closes the meta connection with the given
.Ar NODE .
.It top
If
.Nm
is compiled with libcurses support, this will display live traffic statistics
for all the known nodes, similar to the UNIX
.Xr top 1
command.
See below for more information.
.It pcap
Dump VPN traffic going through the local tinc node in
.Xr pcap-savefile 5
format to standard output,
from where it can be redirected to a file or piped through a program that can parse it directly,
such as
.Xr tcpdump 8 .
.It network Op Ar netname
If
.Ar netname
is given, switch to that network.
Otherwise, display a list of all networks for which configuration files exist.
.It fsck
This will check the configuration files for possible problems,
such as unsafe file permissions, missing executable bit on script,
unknown and obsolete configuration variables, wrong public and/or private keys, and so on.
.Pp
When problems are found, this will be printed on a line with WARNING or ERROR in front of it.
Most problems must be corrected by the user itself, however in some cases (like file permissions and missing public keys),
tinc will ask if it should fix the problem.
.It sign Op Ar filename
Sign a file with the local node's private key.
If no
.Ar filename
is given, the file is read from standard input.
The signed file is written to standard output.
.It verify Ar name Op Ar filename
Check the signature of a file against a node's public key.
The
.Ar name
of the node must be given,
or can be
.Li .
to check against the local node's public key, or
.Li *
to allow a signature from any node whose public key is known.
If no
.Ar filename
is given, the file is read from standard input.
If the verification is successful,
a copy of the input with the signature removed is written to standard output,
and the exit code will be zero.
If the verification failed,
nothing will be written to standard output, and the exit code will be non-zero.
.El
.Sh EXAMPLES
Examples of some commands:
.Bd -literal -offset indent
tinc -n vpn dump graph | circo -Txlib
tinc -n vpn pcap | tcpdump -r -
tinc -n vpn top
.Pp
.Ed
Examples of changing the configuration using
.Nm :
.Bd -literal -offset indent
tinc -n vpn init foo
tinc -n vpn add Subnet 192.168.1.0/24
tinc -n vpn add bar.Address bar.example.com
tinc -n vpn add ConnectTo bar
tinc -n vpn export | gpg --clearsign | mail -s "My config" vpnmaster@example.com
.Ed
.Sh TOP
The top command connects to a running tinc daemon and repeatedly queries its per-node traffic counters.
It displays a list of all the known nodes in the left-most column,
and the amount of bytes and packets read from and sent to each node in the other columns.
By default, the information is updated every second.
The behaviour of the top command can be changed using the following keys:
.Bl -tag
.It Ic s
Change the interval between updates.
After pressing the
.Ic s
key, enter the desired interval in seconds, followed by enter.
Fractional seconds are honored.
Intervals lower than 0.1 seconds are not allowed.
.It Ic c
Toggle between displaying current traffic rates (in packets and bytes per second)
and cumulative traffic (total packets and bytes since the tinc daemon started).
.It Ic n
Sort the list of nodes by name.
.It Ic i
Sort the list of nodes by incoming amount of bytes.
.It Ic I
Sort the list of nodes by incoming amount of packets.
.It Ic o
Sort the list of nodes by outgoing amount of bytes.
.It Ic O
Sort the list of nodes by outgoing amount of packets.
.It Ic t
Sort the list of nodes by sum of incoming and outgoing amount of bytes.
.It Ic T
Sort the list of nodes by sum of incoming and outgoing amount of packets.
.It Ic b
Show amount of traffic in bytes.
.It Ic k
Show amount of traffic in kilobytes.
.It Ic M
Show amount of traffic in megabytes.
.It Ic G
Show amount of traffic in gigabytes.
.It Ic q
Quit.
.El
.Sh BUGS
If you find any bugs, report them to tinc@tinc-vpn.org.
.Sh SEE ALSO
.Xr tincd 8 ,
.Xr tinc.conf 5 ,
.Xr dotty 1 ,
.Xr pcap-savefile 5 ,
.Xr tcpdump 8 ,
.Xr top 1 ,
.Pa http://www.tinc-vpn.org/ ,
.Pa http://www.cabal.org/ .
.Pp
The full documentation for tinc is maintained as a Texinfo manual.
If the info and tinc programs are properly installed at your site,
the command
.Ic info tinc
should give you access to the complete manual.
.Pp
tinc comes with ABSOLUTELY NO WARRANTY.
This is free software, and you are welcome to redistribute it under certain conditions;
see the file COPYING for details.
.Sh AUTHORS
.An "Ivo Timmermans"
.An "Guus Sliepen" Aq guus@tinc-vpn.org
.Pp
And thanks to many others for their contributions to tinc!


NEWS
# Version 1.1pre18             June 27 2021

* Check all Address statements when making outgoing connections.
* Make more variables safe for use in invitations.
* Allow "tinc --force join" to accept all variables sent in an invitation.
* Make sure the stop command works on Windows if tincd is running in the
  foreground.
* Handle DOS line endings in invitation files.
* Double-quote node names in dump graph output.
* Prevent large amounts of UDP probes being sent consecutively.
* Try harder to reconnect with unreachable nodes.
* Generate tinc-up.bat on Windows.
* Fix a possible infinite loop when adding Subnets to a running tincd.
* Allow a tun/tap filedescriptor to be passed through a UNIX socket.
* Use auto-clone tun/tap devices as default on FreeBSD and DragonFlyBSD.

Thanks to Fabian Maurer, Ilia Pavlikhin, Maciej S. Szmigiero, Pacien
Tran-Girard, Aaron Li, Andreas Rammhold, Rosen Penev, Shengjing Zhu, Werner
Schreiber, iczero and leptonyu for their contributions to this version of tinc.

# Version 1.1pre17             October 8 2018

* Prevent oracle attacks in the legacy protocol (CVE-2018-16737,
  CVE-2018-16738).
* Prevent a MITM from forcing a NULL cipher for UDP in the legacy protocol
  (CVE-2018-16758).
* AutoConnect is now enabled by default.
* Per-node network traffic statistics are now shown in the output of "info" and
  "dump nodes" commands.

Thanks to volth and Rafael Sadowski for their contributions to this version of
tinc.

# Version 1.1pre16             June 12 2018

* Fixed building with support for UML sockets.
* Documentation updates and spelling fixes.
* Support for MSS clamping of IP-in-IP packets.
* Fixed parsing of the -b flag.
* Added the ability to set a firemall mark on sockets on Linux.
* Minor improvements to the build system.
* Added a cache of recently seen addresses of peers.
* Add support for --runstatedir to the configure script.
* Fixed linking with libncurses on some distributions.
* Automatically disable PMTUDiscovery when TCPOnly is enabled.
* Fixed removing the tinc service on Windows in some situations.
* Fixed the TAP-Win32 device locking up after waking up from suspend.

Thanks to Todd C. Miller, Etienne Dechamps, Daniel Lublin,
Gjergji Ramku, Mike Sullivan and Oliver Freyermuth for their
contributions to this version of tinc.

# Version 1.1pre15             September 2 2017

* Detect when the machine is resuming from suspension or hibernation.
* When an old PID file is found, check whether the old daemon is still alive.
* Remember scope_id for IPv6 addresses when sending UDP packets to link-local
  addresses.
* Ensure compatibility with OpenSSL 1.1.
* Only log about dropped packets with debug level 5.
* Warn when trying to generate RSA keys less than 2048 bits.
* Use AES256 and SHA256 as the default encryption and digest algorithms.
* Add DeviceType = fd to support tinc on Android without requiring root.
* Support PriorityInheritance for IPv6 packets.
* Fixes for Solaris tun/tap support.
* Add a configurable expiration time for invitations.
* Store invitation data after a successful join.
* Exit gracefully when the tun/tap device is in a bad state.
* Add the LogLevel option.
* AutoConnect now actively tries to heal split networks.

Thanks to Etienne Dechamps, Rafał Leśniak, Sean McVeigh, Vittorio Gambaletta,
Dennis Lan, Pacien Tran-Girard, Roman Savelyev, lemoer and volth for their
contributions to this version of tinc.

# Version 1.1pre14             May 1 2016

* Add tinc.service back.

# Version 1.1pre13             April 30 2016

* Fix BSD tun device support that was broken in 1.1pre12.
* Speed up AutoConnect when there are many host config files present without
  an Address.

# Version 1.1pre12             April 24 2016

* Added a "--syslog" option to force logging to syslog even if running in the
  foreground.
* Fixes and improvements to the DecrementTTL function.
* Improved PMTU discovery and UDP keepalive probes.
* More efficient relaying of UDP packets through intermediate nodes.
* Improved compatibility with newer TAP-Win32 drivers.
* Added support for UPnP.
* Allow tinc to be compiled without LibreSSL or OpenSSL (this drops
  compatibility with nodes running 1.0.x).
* Added a "fsck" command to check the configuration files for problems.
* Tinc "start" now checks whether the daemon really started successfully, and
  displays error messages otherwise.
* Added systemd service files.
* Use the recvmmsg() function if available.
* Support ToS/Diffserv on IPv6 connections.
* Updated support for BSD tun/tap devices.
* Added support for OS X utun interfaces.
* Dropped support for Windows 2000.
* Initial support for generating a tinc-up script from an invitation.
* Many small fixes, documentation updates.

Thanks to Etienne Dechamps, Rafał Leśniak, Vittorio Gambaletta, Martin Weinelt,
Sven-Haegar Koch, Florian Klink, LunnarShadow, Dato Simó, Jo-Philipp Wich,
Jochen Voss, Nathan Stratton Treadway, Pierre Emeriaud, xentec, Samuel
Thibault and Michael Tokarev for their contributions to this version of tinc.

# Version 1.1pre11             December 27 2014

* Added a "network" command to list or switch networks.
* Switched to Ed25519 keys and the ChaCha-Poly1305 cipher for the new protocol.
* AutoConnect is now a boolean option, when enabled tinc always tries to keep
  at least three meta-connections open.
* The new protocol now uses UDP much more often.
* Tinc "del" and "get" commands now return a non-zero exit code when they
  don't find the requested variable.
* Updated documentation.
* Added a "DeviceStandby" option to defer running tinc-up until a working
  connection is made, and which on Windows will also change the network
  interface link status accordingly.
* Tinc now tells the resolver to reload /etc/resolv.conf when it receives
  SIGALRM.
* Improved error messages and event loop handling on Windows.
* LocalDiscovery now uses local address learned from other nodes, and is
  enabled by default.

gui/Tinc.py
#!/usr/bin/python

import string
import socket

REQ_STOP = 0
REQ_RELOAD = 1
REQ_RESTART = 2
REQ_DUMP_NODES = 3
REQ_DUMP_EDGES = 4
REQ_DUMP_SUBNETS = 5
REQ_DUMP_CONNECTIONS = 6
REQ_DUMP_GRAPH = 7
REQ_PURGE = 8
REQ_SET_DEBUG = 9
REQ_RETRY = 10
REQ_CONNECT = 11
REQ_DISCONNECT = 12

ID = 0
ACK = 4
CONTROL = 18

class Node:
	def __init__(self):
		print('New node')

	def __exit__(self):
		print('Deleting node ' + self.name)

	def parse(self, args):
		self.name = args[0]
		self.address = args[2]
		if args[3] != 'port':
			args.insert(3, 'port')
			args.insert(4, '')
		self.port = args[4]
		self.cipher = int(args[6])
		self.digest = int(args[8])
		self.maclength = int(args[10])
		self.compression = int(args[12])
		self.options = int(args[14], 0x10)
		self.status = int(args[16], 0x10)
		self.nexthop = args[18]
		self.via = args[20]
		self.distance = int(args[22])
		self.pmtu = int(args[24])
		self.minmtu = int(args[26])
		self.maxmtu = int(args[28][:-1])

		self.subnets = {}

class Edge:
	def parse(self, args):
		self.fr = args[0]
		self.to = args[2]
		self.address = args[4]
		self.port = args[6]
		self.options = int(args[8], 16)
		self.weight = int(args[10])

class Subnet:
	def parse(self, args):
		if args[0].find('#') >= 0:
			(address, self.weight) = args[0].split('#', 1)
		else:
			self.weight = 10
			address = args[0]

		if address.find('/') >= 0:
			(self.address, self.prefixlen) = address.split('/', 1)
		else:
			self.address = address
			self.prefixlen = '48'

		self.owner = args[2]	

class Connection:
	def parse(self, args):
		self.name = args[0]
		self.address = args[2]
		if args[3] != 'port':
			args.insert(3, 'port')
			args.insert(4, '')
		self.port = args[4]
		self.options = int(args[6], 0x10)
		self.socket = int(args[8])
		self.status = int(args[10], 0x10)
		self.weight = 123

class VPN:
	confdir = '/etc/tinc'
	cookiedir = '/var/run/'

	def connect(self):
		f = open(self.cookiefile)
		cookie = string.split(f.readline())
		f.close()
		s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
		s.connect(('127.0.0.1', int(cookie[1])))
		self.sf = s.makefile()
		s.close()
		hello = string.split(self.sf.readline())
		self.name = hello[1]
		self.sf.write('0 ^' + cookie[0] + ' 17\r\n')
		self.sf.flush()
		resp = string.split(self.sf.readline())
		self.port = cookie[1]
		self.nodes = {}
		self.edges = {}
		self.subnets = {}
		self.connections = {}
		self.refresh()

	def refresh(self):
		self.sf.write('18 3\r\n18 4\r\n18 5\r\n18 6\r\n')
		self.sf.flush()

		for node in self.nodes.values():
			node.visited = False
		for edge in self.edges.values():
			edge.visited = False
		for subnet in self.subnets.values():
			subnet.visited = False
		for connections in self.connections.values():
			connections.visited = False

		while True:
			resp = string.split(self.sf.readline())
			if len(resp) < 2:
				break
			if resp[0] != '18':
				break
			if resp[1] == '3':
				if len(resp) < 3:
					continue
				node = self.nodes.get(resp[2]) or Node()
				node.parse(resp[2:])
				node.visited = True
				self.nodes[resp[2]] = node
			elif resp[1] == '4':
				if len(resp) < 5:
					continue
				edge = self.nodes.get((resp[2], resp[4])) or Edge()
				edge.parse(resp[2:])
				edge.visited = True
				self.edges[(resp[2], resp[4])] = edge
			elif resp[1] == '5':
				if len(resp) < 5:
					continue
				subnet = self.subnets.get((resp[2], resp[4])) or Subnet()
				subnet.parse(resp[2:])
				subnet.visited = True
				self.subnets[(resp[2], resp[4])] = subnet
				self.nodes[subnet.owner].subnets[resp[2]] = subnet
			elif resp[1] == '6':
				if len(resp) < 5:
					break
				connection = self.connections.get((resp[2], resp[4])) or Connection()
				connection.parse(resp[2:])
				connection.visited = True
				self.connections[(resp[2], resp[4])] = connection
			else:
				break

		for key, subnet in self.subnets.items():
			if not subnet.visited:
				del self.subnets[key]

		for key, edge in self.edges.items():
			if not edge.visited:
				del self.edges[key]

		for key, node in self.nodes.items():
			if not node.visited:
				del self.nodes[key]
			else:
				for key, subnet in node.subnets.items():
					if not subnet.visited:
						del node.subnets[key]

		for key, connection in self.connections.items():
			if not connection.visited:
				del self.connections[key]

	def close(self):
		self.sf.close()

	def disconnect(self, name):
		self.sf.write('18 12 ' + name + '\r\n')
		self.sf.flush()
		resp = string.split(self.sf.readline())

	def debug(self, level = -1):
		self.sf.write('18 9 ' + str(level) + '\r\n')
		self.sf.flush()
		resp = string.split(self.sf.readline())
		return int(resp[2])

	def __init__(self, netname = None, controlcookie = None):
		self.tincconf = VPN.confdir + '/'

		if netname:
			self.netname = netname
			self.tincconf += netname + '/'

		self.tincconf += 'tinc.conf'

		if controlcookie is not None:
			self.cookiefile = controlcookie
		else:
			self.cookiefile = VPN.cookiedir + 'tinc.'
			if netname:
				self.cookiefile += netname + '.'
			self.cookiefile += 'cookie'
